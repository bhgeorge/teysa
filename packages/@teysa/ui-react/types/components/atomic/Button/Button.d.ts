import { MouseEvent, ReactNode, TouchEvent } from 'react';
declare type ButtonVariant = 'primary' | 'secondary';
interface ButtonProps {
    children: ReactNode;
    disabled?: boolean;
    href?: string;
    onClick?: (event: MouseEvent | TouchEvent) => void;
    type?: 'button' | 'submit';
    variant?: ButtonVariant;
}
declare const Button: ({ children, disabled, href, onClick, type, variant, }: ButtonProps) => JSX.Element;
export default Button;
