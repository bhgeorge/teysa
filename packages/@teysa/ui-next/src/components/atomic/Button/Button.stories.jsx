// Local
import { Button } from './Button';

export default {
  title: 'Atomic/Button',
  component: Button,
  argTypes: { onClick: { action: 'clicked' } },
};

const Template = props => <Button {...props}>Click me</Button>;

export const Primary = Template.bind({});

export const Secondary = Template.bind({});
Secondary.args = {
  variant: 'secondary',
};

export const AsLink = Template.bind({});
AsLink.args = {
  href: 'https://scryfall.com',
};
