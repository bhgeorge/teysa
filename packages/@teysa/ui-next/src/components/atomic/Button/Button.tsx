// Global
import React, { MouseEvent, ReactNode, TouchEvent } from 'react';
import classNames from 'classnames';
// Components
import { Link } from '../Link/Link';

type ButtonVariant = 'primary' | 'secondary';

interface ButtonProps {
  children: ReactNode;
  disabled?: boolean;
  href?: string;
  onClick?: (event: MouseEvent | TouchEvent) => void;
  type?: 'button' | 'submit';
  variant?: ButtonVariant;
}

const buttonBaseClasses = classNames(
  'border-2',
  'font-bold',
  'font-regular',
  'group',
  'hover:border-theme-interact-hover',
  'hover:underline',
  'inline-flex',
  'items-center',
  'px-2',
  'py-1',
  'rounded-sm',
  'self-start',
  'self-start'
);

const variantClasses: Record<ButtonVariant, string> = {
  primary: classNames(
    buttonBaseClasses,
    'border-theme-interact',
    'bg-theme-interact',
    'text-theme-bg',
    'hover:bg-theme-interact-hover'
  ),
  secondary: classNames(
    buttonBaseClasses,
    'border-theme-interact',
    'hover:text-theme-interact-hover'
  ),
};

export function Button({
  children,
  disabled,
  href,
  onClick,
  type = 'button',
  variant = 'primary',
}: ButtonProps) {
  const cN = classNames(variantClasses[variant]);

  if (href) {
    return (
      <Link href={href} onClick={onClick} className={cN}>
        {children}
      </Link>
    );
  }

  return (
    <button type={type} className={cN} onClick={onClick} disabled={disabled}>
      {children}
    </button>
  );
}
