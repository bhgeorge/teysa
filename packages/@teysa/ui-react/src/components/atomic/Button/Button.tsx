// Global
import React, { MouseEvent, ReactNode, TouchEvent } from 'react';
import classNames from 'classnames';
// Components
import Link from '../Link/Link';

type ButtonVariant = 'primary' | 'secondary';

interface ButtonProps {
  children: ReactNode;
  disabled?: boolean;
  href?: string;
  onClick?: (event: MouseEvent | TouchEvent) => void;
  type?: 'button' | 'submit';
  variant?: ButtonVariant;
}

const buttonBaseClasses = classNames(
  'border-2',
  'group',
  'px-2',
  'py-1',
  'font-regular',
  'rounded-sm',
  'self-start',
  'hover:underline',
  'hover:border-theme-interact-hover',
  'font-bold',
  'inline-flex',
  'self-start',
  'items-center'
);

const variantClasses: Record<ButtonVariant, string> = {
  primary: classNames(
    buttonBaseClasses,
    'border-theme-interact',
    'bg-theme-interact',
    'text-theme-bg',
    'hover:bg-theme-interact-hover'
  ),
  secondary: classNames(
    buttonBaseClasses,
    'border-theme-interact',
    'hover:text-theme-interact-hover'
  ),
};

const Button = ({
  children,
  disabled,
  href,
  onClick,
  type = 'button',
  variant = 'primary',
}: ButtonProps) => {
  const cN = classNames(variantClasses[variant]);

  if (href) {
    return (
      <Link href={href} onClick={onClick} className={classNames(cN, 'inline-flex')}>
        <span className="inline-flex items-center justify-between relative z-20">{children}</span>
        <span
          style={{ height: '200%' }}
          className="absolute top-0 left-0 w-full transform translate-y-0 transition-transform group-hover:-translate-y-1/2 interact-gradient"
        />
      </Link>
    );
  }

  return (
    <button type={type} className={cN} onClick={onClick} disabled={disabled}>
      {children}
    </button>
  );
};

export default Button;
