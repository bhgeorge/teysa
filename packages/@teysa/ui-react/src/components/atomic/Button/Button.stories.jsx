// Local
import Button from './Button';

export default {
  title: 'Atomic/Button',
  component: Button,
};

const Template = (props) => <Button {...props} />;

export const Default = Template.bind({});
Default.args = {
  text: 'Click me',
};
