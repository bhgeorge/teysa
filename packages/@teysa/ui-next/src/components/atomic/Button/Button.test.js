// Global
import React from 'react';
import { render, fireEvent } from '@testing-library/react';
// Local
import { Button } from './Button';

const createComponent = props => render(<Button {...props}>Click me</Button>);

describe('it has an href', () => {
  it('uses the Link component', () => {
    const component = createComponent({ href: '/test' });
    expect(component.container.firstChild.tagName).toBe('A');
  });
});

describe('it is a button', () => {
  it('displays as a <button>', () => {
    const CLICK_VAL = 'clicked';
    let clickVal = null;
    const component = createComponent({
      onClick: () => {
        clickVal = CLICK_VAL;
      },
    });

    const button = component.container.firstChild;
    expect(button).toMatchSnapshot();
    expect(button.getAttribute('type')).toBe('button');
    // Test onClick
    fireEvent.click(button);
    expect(clickVal).toBe(CLICK_VAL);
  });

  it('displays as submit', () => {
    const component = createComponent({ type: 'submit' });
    expect(component.container.firstChild.getAttribute('type')).toBe('submit');
  });

  it('displays a secondary variant', () => {
    const component = createComponent({ variant: 'secondary' });
    expect(component.container.firstChild).toMatchSnapshot();
  });
});
