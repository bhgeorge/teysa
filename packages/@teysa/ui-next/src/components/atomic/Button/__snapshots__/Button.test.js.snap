// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it is a button displays a secondary variant 1`] = `
<button
  class="border-2 group px-2 py-1 font-regular rounded-sm self-start hover:underline hover:border-theme-interact-hover font-bold inline-flex self-start items-center border-theme-interact hover:text-theme-interact-hover"
  type="button"
>
  Click me
</button>
`;

exports[`it is a button displays as a <button> 1`] = `
<button
  class="border-2 group px-2 py-1 font-regular rounded-sm self-start hover:underline hover:border-theme-interact-hover font-bold inline-flex self-start items-center border-theme-interact bg-theme-interact text-theme-bg hover:bg-theme-interact-hover"
  type="button"
>
  Click me
</button>
`;
